src/storage-controller/src/collection_mgmt.rs

```diff
@@ -499,8 +549,19 @@ where
                     }
                 }
 
+                _it_changed = self.read_only.changed() => {
+                    // We can now write, attempt to do that right away, even if
+                    // our `to_write` is empty. This way, we might learn that
+                    // there is something in persist that we have to retract.
+                    let  _ = self.write_to_persist(vec![]).await?;
+                }
+
                 // If we haven't received any updates, then we'll move the upper forward.
                 _ = self.upper_tick_interval.tick() => {
+                    if *self.read_only.borrow() {
+                        // Not bumping uppers while in read-only mode.
+                        continue;
+                    }
                     let _ = self.tick_upper().await?;
```

Comment:
Is there a reasonable place to assert that we don't do writes in read-only mode? Before each `self.write_handle.compare_and_append` call maybe? Or within `PersistMonotonicWriteWorker`?

Reply:
I'll add an assert before each `compare_and_append` (there's only two!). I considered teaching `PersistMonotonicWriteWorker` about read-only mode but then thought that it's not worth the effort, and `CollectionManager` is largely the thing through which writes are facilitated.